engine/src/flutter/shell/platform/darwin/macos/framework/Source/ResizeSynchronizer.swift

```diff
@@ -0,0 +1,151 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import CoreGraphics
+import Foundation
+
+/// Coordinates Flutter view content updates with macOS window resizing.
+///
+/// When a native window containing a Flutter view is resized, the platform's window bounds can
+/// change before Flutter has been able to render a new frame matching those new dimensions. This
+/// asynchronicity can lead to undesirable visual effects such as:
+///   - Content temporarily appearing at the old size within the new, larger bounds.
+///   - Black "letterboxing" artifacts rendered at view edges.
+///   - A perception of lag in content resizing.
+///
+/// `ResizeSynchronizer` mitigates these issues by introducing a controlled blocking mechanism
+/// during the resize lifecycle:
+///
+/// 1.  Initiation (Platform Thread): When a native window resize is detected (e.g., by a window
+///     delegate), the platform thread calls `beginResize(forSize:notify:onTimeout:)` with the new
+///     target dimensions. This method blocks the calling platform thread, pausing the completion of
+///     the native resize event. While blocked, it continuously polls for engine messages to process
+///     relevant updates. The `notify` closure is invoked just before blocking begins.
+///
+/// 2.  Commit (Raster Thread): As the engine renders frames, the raster thread, upon presenting a
+///     new frame, calls `performCommit(forSize:afterDelay:notify:)`, providing the size of the
+///     frame just rendered.
+///
+/// 3.  Unblocking: If the size of the committed frame matches the target size that `beginResize()`
+///     is waiting for, `beginResize()` unblocks. This allows the native resize operation to
+///     complete, now that Flutter's content is synchronized with the new window dimensions.
+///
+/// Safeguards:
+///   - Timeout: `beginResize()` includes a timeout mechanism to prevent indefinite blocking if a
+///     matching frame isn't committed in a timely manner. An optional `onTimeout` closure can be
+///     provided to handle this event.
+///   - Shutdown: The synchronization can be cleanly interrupted by calling `shutDown()`, which will
+///     also unblock any pending `beginResize()` call. After shutdown, `beginResize()` will no
+///     longer block.
```

Comment:
Seems to me after shutdown, we stop synchronizing since `isShuttingDown` never becomes false again. May be worth adding to this comment?

Reply:
Oh or maybe that's what you meant by "will no longer block"

Reply:
Ah yep that's exactly what I mean. 

i.e. no need to block any longer since when shutting down, the app is going away so there's no more need to synchronise window updates with flutter content updates.